Adds new example NN API. Removes old.

ry opened the pull request:
fixes #259

piscisaureus commented on example.js:
```diff
@@ -0,0 +1,56 @@
+        .linear("L3", params, 10)
+        // Using the logits, calculate a classification loss between the
+        // labels. Labels's shape is [128]. This value is returned, and will be
+        // backpropagated though.
```
through.

piscisaureus commented on src/experiment.ts:
```diff
@@ -0,0 +1,144 @@
+    return;
+  }
+  const promise = printHelper(...args);
+  // All this is to make sure that log items are in the order they came.
```
This is very complicated. I would suggest the following:

* lastPrint is *always* a promise. You can make an empty resolved promise with `Promise.resolve()`.
* Unconditionally do the following: `lastPrint = lastPrint.then(() => printHelper(...args))`

ry replied:
Done.

piscisaureus commented on src/experiment.ts:
```diff
@@ -0,0 +1,144 @@
+    if (typeof arg === "string") {
+      return arg;
+    } else if (typeof arg === "number") {
+      return String(arg);
```
This conversion is unnecessary, and this if...elseif...else block is missing cases for certain cases like booleans, null etc.
I would suggest to check for tensors explicitly, and otherwise just return the input.
```
if (arg instanceof Tensor) {
  return format.toString(arg.shape, await arg.data());
} else {
  return arg;
}
```

ry replied:
done

piscisaureus commented on src/experiment.ts:
```diff
@@ -0,0 +1,144 @@
+  // Drop prints if we get too backed up.
+  if (printWaitCount > 100) {
+    console.log("drop print");
+    await delay(0);
```
Why this delay?

ry replied:
Cruft from earlier debugging. Removed.

piscisaureus commented on src/experiment.ts:
```diff
@@ -0,0 +1,144 @@
+      optimizer(opts, this.currentParams, grads);
+    });
+
+    await print("step", this.step, "loss", loss);
```
The `await` statement here defeats the purpose of throttling in your `print` function.
So maybe it may not make a lot of sense to address my comments above, because why do it at all?

ry replied:
Made async.

ry commented:
by snapshotting do you mean checkpointing? that will be done every N steps or every so many minutes - not every step.

I think the printing backoff isn't working because TF's tensor.data() is actually synchronous, same for DL CPU. It probably only happens when using webgl. However, given that we can't test it, I've removed that code. 

I've also updated the maxSteps to 3k.

Another important point is that this network hasn't been tested on GPU yet - while nn_example has.. Ideally this example would default to using GPUs without additional code.
